File: src/Ncp.Admin.Domain/AggregatesModel/CustomerAggregate/Customer.cs
using Ncp.Admin.Domain;
using Ncp.Admin.Domain.AggregatesModel.CustomerSourceAggregate;
using Ncp.Admin.Domain.AggregatesModel.IndustryAggregate;
using Ncp.Admin.Domain.AggregatesModel.UserAggregate;
using Ncp.Admin.Domain.DomainEvents.CustomerEvents;

namespace Ncp.Admin.Domain.AggregatesModel.CustomerAggregate;

/// <summary>
/// 客户 ID（强类型）
/// </summary>
public partial record CustomerId : IGuidStronglyTypedId;

/// <summary>
/// 客户聚合根：档案、联系人、行业、公海归属等
/// </summary>
public class Customer : Entity<CustomerId>, IAggregateRoot
{
    /// <summary>
    /// EF/序列化用
    /// </summary>
    protected Customer() { }

    /// <summary>
    /// 联系人列表
    /// </summary>
    public virtual ICollection<CustomerContact> Contacts { get; } = [];

    /// <summary>
    /// 所属行业关联列表
    /// </summary>
    public virtual ICollection<CustomerIndustry> Industries { get; } = [];

    /// <summary>
    /// 负责人用户 ID
    /// </summary>
    public UserId? OwnerId { get; private set; }

    /// <summary>
    /// 客户来源 ID（主数据，必填）
    /// </summary>
    public CustomerSourceId CustomerSourceId { get; private set; } = default!;

    /// <summary>
    /// 客户来源名称（冗余，便于列表/展示）
    /// </summary>
    public string CustomerSourceName { get; private set; } = string.Empty;

    /// <summary>
    /// 是否作废客户
    /// </summary>
    public bool IsVoided { get; private set; }

    /// <summary>
    /// 客户全称
    /// </summary>
    public string FullName { get; private set; } = string.Empty;

    /// <summary>
    /// 客户简称
    /// </summary>
    public string ShortName { get; private set; } = string.Empty;

    /// <summary>
    /// 公司性质
    /// </summary>
    public string Nature { get; private set; } = string.Empty;

    /// <summary>
    /// 省区域码
    /// </summary>
    public string ProvinceCode { get; private set; } = string.Empty;

    /// <summary>
    /// 市区域码
    /// </summary>
    public string CityCode { get; private set; } = string.Empty;

    /// <summary>
    /// 区/县区域码
    /// </summary>
    public string DistrictCode { get; private set; } = string.Empty;

    /// <summary>
    /// 省名称（冗余，便于列表/展示）
    /// </summary>
    public string ProvinceName { get; private set; } = string.Empty;

    /// <summary>
    /// 市名称（冗余，便于列表/展示）
    /// </summary>
    public string CityName { get; private set; } = string.Empty;

    /// <summary>
    /// 区/县名称（冗余，便于列表/展示）
    /// </summary>
    public string DistrictName { get; private set; } = string.Empty;

    /// <summary>
    /// 电话区域-省区域码
    /// </summary>
    public string PhoneProvinceCode { get; private set; } = string.Empty;

    /// <summary>
    /// 电话区域-市区域码
    /// </summary>
    public string PhoneCityCode { get; private set; } = string.Empty;

    /// <summary>
    /// 电话区域-区/县区域码
    /// </summary>
    public string PhoneDistrictCode { get; private set; } = string.Empty;

    /// <summary>
    /// 电话区域-省/市/区名称（冗余）
    /// </summary>
    public string PhoneProvinceName { get; private set; } = string.Empty;

    public string PhoneCityName { get; private set; } = string.Empty;

    public string PhoneDistrictName { get; private set; } = string.Empty;

    /// <summary>
    /// 咨询内容（公海录入）
    /// </summary>
    public string ConsultationContent { get; private set; } = string.Empty;

    /// <summary>
    /// 客户覆盖区域
    /// </summary>
    public string CoverRegion { get; private set; } = string.Empty;

    /// <summary>
    /// 公司注册地址
    /// </summary>
    public string RegisterAddress { get; private set; } = string.Empty;

    /// <summary>
    /// 主联系人姓名
    /// </summary>
    public string MainContactName { get; private set; } = string.Empty;

    /// <summary>
    /// 主联系方式
    /// </summary>
    public string MainContactPhone { get; private set; } = string.Empty;

    /// <summary>
    /// 微信添加情况
    /// </summary>
    public string WechatStatus { get; private set; } = string.Empty;

    /// <summary>
    /// 备注
    /// </summary>
    public string Remark { get; private set; } = string.Empty;

    /// <summary>
    /// 联系人QQ（公海/线索）
    /// </summary>
    public string ContactQq { get; private set; } = string.Empty;

    /// <summary>
    /// 联系人微信（公海/线索，与 WechatStatus 区分：此处存微信号等）
    /// </summary>
    public string ContactWechat { get; private set; } = string.Empty;

    /// <summary>
    /// 是否大客户
    /// </summary>
    public bool IsKeyAccount { get; private set; }

    /// <summary>
    /// 是否隐藏
    /// </summary>
    public bool IsHidden { get; private set; }

    /// <summary>
    /// 合并标识
    /// </summary>
    public bool CombineFlag { get; private set; }

    /// <summary>
    /// 是否在公海
    /// </summary>
    public bool IsInSea { get; private set; }

    /// <summary>
    /// 释放到公海时间
    /// </summary>
    public DateTimeOffset? ReleasedToSeaAt { get; private set; }

    /// <summary>
    /// 创建人用户 ID
    /// </summary>
    public UserId CreatorId { get; private set; } = default!;

    /// <summary>
    /// 创建人姓名（冗余，便于列表/展示）
    /// </summary>
    public string CreatorName { get; private set; } = string.Empty;

    /// <summary>
    /// 领用人/负责人姓名（冗余，OwnerId 对应姓名，便于列表/展示）
    /// </summary>
    public string OwnerName { get; private set; } = string.Empty;

    /// <summary>
    /// 领用时间（从公海领用时写入，释放回公海时清空）
    /// </summary>
    public DateTimeOffset? ClaimedAt { get; private set; }

    /// <summary>
    /// 创建时间
    /// </summary>
    public DateTimeOffset CreatedAt { get; init; }

    /// <summary>
    /// 最后更新时间
    /// </summary>
    public UpdateTime UpdateTime { get; private set; } = new UpdateTime(DateTimeOffset.UtcNow);

    /// <summary>
    /// 创建客户；未传负责人则视为公海客户
    /// </summary>
    public Customer(
        UserId? ownerId,
        CustomerSourceId customerSourceId,
        string customerSourceName,
        string fullName,
        string shortName,
        string nature,
        string provinceCode,
        string cityCode,
        string districtCode,
        string provinceName,
        string cityName,
        string districtName,
        string phoneProvinceCode,
        string phoneCityCode,
        string phoneDistrictCode,
        string phoneProvinceName,
        string phoneCityName,
        string phoneDistrictName,
        string consultationContent,
        string coverRegion,
        string registerAddress,
        string mainContactName,
        string mainContactPhone,
        string contactQq,
        string contactWechat,
        string wechatStatus,
        string remark,
        bool isKeyAccount,
        UserId creatorId,
        string creatorName)
    {
        OwnerId = ownerId;
        CustomerSourceId = customerSourceId;
        CustomerSourceName = customerSourceName;
        IsVoided = false;
        FullName = fullName;
        ShortName = shortName;
        Nature = nature;
        ProvinceCode = provinceCode;
        CityCode = cityCode;
        DistrictCode = districtCode;
        ProvinceName = provinceName;
        CityName = cityName;
        DistrictName = districtName;
        PhoneProvinceCode = phoneProvinceCode;
        PhoneCityCode = phoneCityCode;
        PhoneDistrictCode = phoneDistrictCode;
        PhoneProvinceName = phoneProvinceName;
        PhoneCityName = phoneCityName;
        PhoneDistrictName = phoneDistrictName;
        ConsultationContent = consultationContent;
        CoverRegion = coverRegion;
        RegisterAddress = registerAddress;
        MainContactName = mainContactName;
        MainContactPhone = mainContactPhone;
        ContactQq = contactQq;
        ContactWechat = contactWechat;
        WechatStatus = wechatStatus;
        Remark = remark;
        IsKeyAccount = isKeyAccount;
        CreatorId = creatorId;
        CreatorName = creatorName;
        OwnerName = string.Empty;
        ClaimedAt = null;
        IsInSea = ownerId == null;
        CreatedAt = DateTimeOffset.UtcNow;
        AddDomainEvent(new CustomerCreatedDomainEvent(this));
    }

    /// <summary>
    /// 公海录入创建客户；仅包含公海录入表单字段，无客户名称/负责人/部门，其余属性为默认值。
    /// </summary>
    public Customer(
        CustomerSourceId customerSourceId,
        string customerSourceName,
        string mainContactName,
        string mainContactPhone,
        string contactQq,
        string contactWechat,
        string phoneProvinceCode,
        string phoneCityCode,
        string phoneDistrictCode,
        string phoneProvinceName,
        string phoneCityName,
        string phoneDistrictName,
        string provinceCode,
        string cityCode,
        string districtCode,
        string provinceName,
        string cityName,
        string districtName,
        string consultationContent,
        UserId creatorId,
        string creatorName)
    {
        OwnerId = null;
        CustomerSourceId = customerSourceId;
        CustomerSourceName = customerSourceName;
        IsVoided = false;
        FullName = string.Empty;
        ShortName = string.Empty;
        Nature = string.Empty;
        ProvinceCode = provinceCode;
        CityCode = cityCode;
        DistrictCode = districtCode;
        ProvinceName = provinceName;
        CityName = cityName;
        DistrictName = districtName;
        PhoneProvinceCode = phoneProvinceCode;
        PhoneCityCode = phoneCityCode;
        PhoneDistrictCode = phoneDistrictCode;
        PhoneProvinceName = phoneProvinceName;
        PhoneCityName = phoneCityName;
        PhoneDistrictName = phoneDistrictName;
        ConsultationContent = consultationContent;
        CoverRegion = string.Empty;
        RegisterAddress = string.Empty;
        MainContactName = mainContactName;
        MainContactPhone = mainContactPhone;
        ContactQq = contactQq;
        ContactWechat = contactWechat;
        WechatStatus = string.Empty;
        Remark = string.Empty;
        IsKeyAccount = false;
        CreatorId = creatorId;
        CreatorName = creatorName;
        OwnerName = string.Empty;
        ClaimedAt = null;
        IsInSea = true;
        CreatedAt = DateTimeOffset.UtcNow;
        AddDomainEvent(new CustomerCreatedDomainEvent(this));
    }

    /// <summary>
    /// 设置所属行业（创建后由命令处理器调用）
    /// </summary>
    public void SetIndustries(IEnumerable<IndustryId> industryIds)
    {
        Industries.Clear();
        foreach (var id in industryIds)
            Industries.Add(CustomerIndustry.Create(Id, id));
    }

    /// <summary>
    /// 更新客户档案；公海客户需先领用后再改
    /// </summary>
    public void Update(
        UserId? ownerId,
        CustomerSourceId customerSourceId,
        string customerSourceName,
        string fullName,
        string shortName,
        string nature,
        string provinceCode,
        string cityCode,
        string districtCode,
        string provinceName,
        string cityName,
        string districtName,
        string phoneProvinceCode,
        string phoneCityCode,
        string phoneDistrictCode,
        string phoneProvinceName,
        string phoneCityName,
        string phoneDistrictName,
        string consultationContent,
        string coverRegion,
        string registerAddress,
        string mainContactName,
        string mainContactPhone,
        string contactQq,
        string contactWechat,
        string wechatStatus,
        string remark,
        bool isKeyAccount,
        bool isHidden,
        IEnumerable<IndustryId>? industryIds = null)
    {
        if (IsInSea)
            throw new KnownException("公海客户需先领用后再修改", ErrorCodes.CustomerNotInSea);
        OwnerId = ownerId;
        CustomerSourceId = customerSourceId;
        CustomerSourceName = customerSourceName;
        FullName = fullName;
        ShortName = shortName;
        Nature = nature;
        ProvinceCode = provinceCode;
        CityCode = cityCode;
        DistrictCode = districtCode;
        ProvinceName = provinceName;
        CityName = cityName;
        DistrictName = districtName;
        PhoneProvinceCode = phoneProvinceCode;
        PhoneCityCode = phoneCityCode;
        PhoneDistrictCode = phoneDistrictCode;
        PhoneProvinceName = phoneProvinceName;
        PhoneCityName = phoneCityName;
        PhoneDistrictName = phoneDistrictName;
        ConsultationContent = consultationContent;
        CoverRegion = coverRegion;
        RegisterAddress = registerAddress;
        MainContactName = mainContactName;
        MainContactPhone = mainContactPhone;
        ContactQq = contactQq;
        ContactWechat = contactWechat;
        WechatStatus = wechatStatus;
        Remark = remark;
        IsKeyAccount = isKeyAccount;
        IsHidden = isHidden;
        UpdateTime = new UpdateTime(DateTimeOffset.UtcNow);
        if (industryIds != null)
        {
            Industries.Clear();
            foreach (var id in industryIds)
                Industries.Add(CustomerIndustry.Create(Id, id));
        }
        AddDomainEvent(new CustomerUpdatedDomainEvent(this));
    }

    /// <summary>
    /// 公海客户档案更新（仅当 IsInSea 时可调用，不变更负责人/部门）
    /// </summary>
    public void UpdateWhenInSea(
        CustomerSourceId customerSourceId,
        string customerSourceName,
        string shortName,
        string nature,
        string provinceCode,
        string cityCode,
        string districtCode,
        string provinceName,
        string cityName,
        string districtName,
        string phoneProvinceCode,
        string phoneCityCode,
        string phoneDistrictCode,
        string phoneProvinceName,
        string phoneCityName,
        string phoneDistrictName,
        string consultationContent,
        string coverRegion,
        string registerAddress,
        string mainContactName,
        string mainContactPhone,
        string contactQq,
        string contactWechat,
        string wechatStatus,
        string remark,
        bool isKeyAccount,
        IEnumerable<IndustryId>? industryIds = null)
    {
        if (!IsInSea)
            throw new KnownException("仅公海客户可在此处修改", ErrorCodes.CustomerNotInSea);
        CustomerSourceId = customerSourceId;
        CustomerSourceName = customerSourceName;
        ShortName = shortName;
        Nature = nature;
        ProvinceCode = provinceCode;
        CityCode = cityCode;
        DistrictCode = districtCode;
        ProvinceName = provinceName;
        CityName = cityName;
        DistrictName = districtName;
        PhoneProvinceCode = phoneProvinceCode;
        PhoneCityCode = phoneCityCode;
        PhoneDistrictCode = phoneDistrictCode;
        PhoneProvinceName = phoneProvinceName;
        PhoneCityName = phoneCityName;
        PhoneDistrictName = phoneDistrictName;
        ConsultationContent = consultationContent;
        CoverRegion = coverRegion;
        RegisterAddress = registerAddress;
        MainContactName = mainContactName;
        MainContactPhone = mainContactPhone;
        ContactQq = contactQq;
        ContactWechat = contactWechat;
        WechatStatus = wechatStatus;
        Remark = remark;
        IsKeyAccount = isKeyAccount;
        UpdateTime = new UpdateTime(DateTimeOffset.UtcNow);
        if (industryIds != null)
        {
            Industries.Clear();
            foreach (var id in industryIds)
                Industries.Add(CustomerIndustry.Create(Id, id));
        }
        AddDomainEvent(new CustomerUpdatedDomainEvent(this));
    }

    /// <summary>
    /// 作废（仅公海客户可作废）
    /// </summary>
    public void Void()
    {
        if (!IsInSea)
            throw new KnownException("仅公海客户可作废", ErrorCodes.CustomerNotInSea);
        IsVoided = true;
        UpdateTime = new UpdateTime(DateTimeOffset.UtcNow);
    }

    /// <summary>
    /// 释放到公海
    /// </summary>
    public void ReleaseToSea()
    {
        if (IsInSea)
            return;
        OwnerId = null;
        OwnerName = string.Empty;
        ClaimedAt = null;
        IsInSea = true;
        ReleasedToSeaAt = DateTimeOffset.UtcNow;
        UpdateTime = new UpdateTime(DateTimeOffset.UtcNow);
        AddDomainEvent(new CustomerReleasedToSeaDomainEvent(this));
    }

    /// <summary>
    /// 从公海领用（仅公海客户可领用）
    /// </summary>
    public void ClaimFromSea(UserId ownerId, string? ownerName)
    {
        if (!IsInSea)
            throw new KnownException("仅公海客户可领用", ErrorCodes.CustomerNotInSea);
        OwnerId = ownerId;
        OwnerName = ownerName ?? string.Empty;
        ClaimedAt = DateTimeOffset.UtcNow;
        IsInSea = false;
        ReleasedToSeaAt = null;
        UpdateTime = new UpdateTime(DateTimeOffset.UtcNow);
        AddDomainEvent(new CustomerClaimedFromSeaDomainEvent(this));
    }

    /// <summary>
    /// 添加联系人
    /// </summary>
    public CustomerContactId AddContact(
        string name,
        string contactType,
        int? gender,
        DateTime? birthday,
        string position,
        string mobile,
        string phone,
        string email,
        bool isPrimary)
    {
        var contact = CustomerContact.Create(Id, name, contactType, gender, birthday, position, mobile, phone, email, isPrimary);
        Contacts.Add(contact);
        if (isPrimary)
        {
            foreach (var c in Contacts.Where(c => c != contact))
                c.Update(c.Name, c.ContactType, c.Gender, c.Birthday, c.Position, c.Mobile, c.Phone, c.Email, false);
        }
        AddDomainEvent(new CustomerContactAddedDomainEvent(this, contact));
        return contact.Id;
    }

    /// <summary>
    /// 更新指定联系人
    /// </summary>
    public void UpdateContact(
        CustomerContactId contactId,
        string name,
        string contactType,
        int? gender,
        DateTime? birthday,
        string position,
        string mobile,
        string phone,
        string email,
        bool isPrimary)
    {
        var contact = Contacts.FirstOrDefault(c => c.Id == contactId)
            ?? throw new KnownException("未找到客户联系人", ErrorCodes.CustomerContactNotFound);
        contact.Update(name, contactType, gender, birthday, position, mobile, phone, email, isPrimary);
        if (isPrimary)
        {
            foreach (var c in Contacts.Where(c => c != contact))
                c.Update(c.Name, c.ContactType, c.Gender, c.Birthday, c.Position, c.Mobile, c.Phone, c.Email, false);
        }
        UpdateTime = new UpdateTime(DateTimeOffset.UtcNow);
        AddDomainEvent(new CustomerContactUpdatedDomainEvent(this, contact));
    }

    /// <summary>
    /// 移除指定联系人
    /// </summary>
    public void RemoveContact(CustomerContactId contactId)
    {
        var contact = Contacts.FirstOrDefault(c => c.Id == contactId)
            ?? throw new KnownException("未找到客户联系人", ErrorCodes.CustomerContactNotFound);
        AddDomainEvent(new CustomerContactRemovedDomainEvent(this, contact.Id));
        Contacts.Remove(contact);
        UpdateTime = new UpdateTime(DateTimeOffset.UtcNow);
    }
}

File: src/frontend/apps/admin-antd/src/api/system/customer.ts
import type { Recordable } from '@vben/types';

import { requestClient } from '#/api/request';

export namespace CustomerApi {
  export interface CustomerItem {
    id: string;
    ownerId?: string;
    ownerDeptId?: string;
    ownerDeptName?: string;
    customerSourceId: string;
    customerSourceName: string;
    fullName: string;
    shortName?: string;
    /** 客户状态枚举值：0 跟进中-暂不明了 1 跟进中-有意向 2 合作中客户 3 曾合作客户 */
    status?: number;
    /** 公司性质枚举值：0 个体 1 民营 2 国企或央企 3 外企 4 其他 5 终端客户 6 运营商 */
    nature?: number;
    provinceCode?: string;
    cityCode?: string;
    districtCode?: string;
    provinceName?: string;
    cityName?: string;
    districtName?: string;
    phoneProvinceCode?: string;
    phoneCityCode?: string;
    phoneDistrictCode?: string;
    phoneProvinceName?: string;
    phoneCityName?: string;
    phoneDistrictName?: string;
    consultationContent?: string;
    contactQq?: string;
    contactWechat?: string;
    coverRegion?: string;
    registerAddress?: string;
    /** 员工数量 */
    employeeCount?: number;
    /** 营业执照（路径或 URL） */
    businessLicense?: string;
    mainContactName?: string;
    mainContactPhone?: string;
    wechatStatus?: string;
    remark?: string;
    isKeyAccount: boolean;
    isHidden: boolean;
    combineFlag: boolean;
    isInSea: boolean;
    releasedToSeaAt?: string;
    creatorId: string;
    creatorName?: string;
    ownerName?: string;
    claimedAt?: string;
    /** 公海作废标记 */
    isVoided?: boolean;
    createdAt: string;
    contactCount: number;
    industryIds: string[];
  }

  export interface CustomerContactItem {
    id: string;
    name: string;
    contactType?: string;
    /** 性别（必填） */
    gender: number;
    /** 生日 ISO 字符串（必填） */
    birthday: string;
    position?: string;
    mobile?: string;
    phone?: string;
    email?: string;
    qq?: string;
    wechat?: string;
    isWechatAdded?: boolean;
    isPrimary: boolean;
  }

  export interface CustomerContactRecordItem {
    id: string;
    recordAt: string;
    /** 1电话 2出差 3微信 4其他 */
    recordType: number;
    title?: string;
    content: string;
    nextVisitAt?: string;
    /** 0待选择 1有效联系 2无效联系 */
    status: number;
    customerContactIds?: string[];
    ownerId?: string;
    ownerName?: string;
    ownerDeptId?: string;
    ownerDeptName?: string;
    creatorId?: string;
    /** 录入人姓名（接口若返回则展示） */
    creatorName?: string;
    createdAt?: string;
    modifierId?: string;
    modifiedAt?: string;
    remark?: string;
    reminderIntervalDays?: number;
    reminderCount?: number;
    filePath?: string;
    customerAddress?: string;
    visitAddress?: string;
    /** @deprecated 使用 status */
    statusId?: number;
    /** @deprecated 使用 ownerName */
    recorderName?: string;
  }

  export interface CustomerDetail extends CustomerItem {
    contacts: CustomerContactItem[];
    contactRecords: CustomerContactRecordItem[];
  }

  export interface CustomerSearchItem {
    id: string;
    fullName: string;
    shortName?: string;
    mainContactPhone?: string;
  }
}

export interface GetCustomerListParams extends Recordable<any> {
  pageIndex?: number;
  pageSize?: number;
  fullName?: string;
  ownerId?: string;
}

async function getCustomerList(params: GetCustomerListParams) {
  const res = await requestClient.get<{
    items: CustomerApi.CustomerItem[];
    total: number;
  }>('/customers', { params });
  return res;
}

async function getCustomer(id: string) {
  return requestClient.get<CustomerApi.CustomerDetail>(`/customers/${id}`);
}

async function createCustomer(
  data: Recordable<any>,
  options?: { idempotencyKey?: string },
) {
  const headers: Record<string, string> = {};
  if (options?.idempotencyKey)
    headers['Idempotency-Key'] = options.idempotencyKey;
  return requestClient.post<{ id: string }>('/customers', data, {
    ...(Object.keys(headers).length ? { headers } : {}),
  });
}

/** 公海录入专用：创建公海客户（无需客户名称、无负责人） */
async function createSeaCustomer(data: Recordable<any>) {
  return requestClient.post<{ id: string }>('/customers/sea', data);
}

/** 公海录入专用：检查联系方式是否重复（电话/QQ/微信） */
async function checkSeaCustomerDuplicateContacts(data: {
  mainContactPhone?: string;
  contactQq?: string;
  contactWechat?: string;
}) {
  return requestClient.post<{
    items: Array<{
      customerId: string;
      customerName: string;
      customerSourceName: string;
      ownerName: string;
      duplicatePhones: string[];
      duplicateQqs: string[];
      duplicateWechats: string[];
    }>;
  }>('/customers/sea/check-duplicate-contacts', {
    mainContactPhone: data.mainContactPhone ?? '',
    contactQq: data.contactQq ?? '',
    contactWechat: data.contactWechat ?? '',
  });
}

async function updateCustomer(id: string, data: Recordable<any>) {
  return requestClient.put(`/customers/${id}`, data);
}

async function getCustomerSearch(params: Recordable<any>) {
  const res = await requestClient.get<{
    items: CustomerApi.CustomerSearchItem[];
    total: number;
  }>('/customers/search', { params });
  return res;
}

async function getCustomerShares(id: string) {
  return requestClient.get<{ sharedToUserIds: string[] }>(`/customers/${id}/shares`);
}

async function shareCustomer(id: string, sharedToUserIds: string[]) {
  return requestClient.post(`/customers/${id}/share`, { id, sharedToUserIds });
}

async function unshareCustomer(id: string, sharedToUserIds: string[]) {
  return requestClient.delete(`/customers/${id}/share`, { data: { id, sharedToUserIds } as any });
}

async function releaseCustomerToSea(id: string) {
  return requestClient.post(`/customers/${id}/release-to-sea`, {});
}

async function claimCustomerFromSea(id: string) {
  return requestClient.post(`/customers/${id}/claim`, {});
}

async function updateSeaCustomer(id: string, data: Recordable<any>) {
  return requestClient.put(`/customers/${id}/sea`, data);
}

async function updateSeaCustomerConsultation(id: string, consultationContent: string) {
  return requestClient.put(`/customers/${id}/sea/consultation`, {
    id,
    consultationContent,
  });
}

export interface VoidCustomerWorkflowResponse {
  workflowInstanceId: string;
  title: string;
}

async function voidCustomer(id: string, routingRoleId?: string) {
  const body = routingRoleId ? { routingRoleId } : {};
  return requestClient.post<VoidCustomerWorkflowResponse>(`/customers/${id}/void`, body);
}

async function deleteCustomer(id: string) {
  return requestClient.delete(`/customers/${id}`);
}

async function addCustomerContact(customerId: string, data: Recordable<any>) {
  return requestClient.post<{ id: string }>(`/customers/${customerId}/contacts`, data);
}

async function updateCustomerContact(
  customerId: string,
  contactId: string,
  data: Recordable<any>,
) {
  return requestClient.put(`/customers/${customerId}/contacts/${contactId}`, data);
}

async function removeCustomerContact(customerId: string, contactId: string) {
  return requestClient.delete(`/customers/${customerId}/contacts/${contactId}`);
}

async function addCustomerContactRecord(customerId: string, data: Recordable<any>) {
  return requestClient.post<{ id: string }>(`/customers/${customerId}/contact-records`, data);
}

async function updateCustomerContactRecord(
  customerId: string,
  recordId: string,
  data: Recordable<any>,
) {
  return requestClient.put(`/customers/${customerId}/contact-records/${recordId}`, data);
}

async function removeCustomerContactRecord(customerId: string, recordId: string) {
  return requestClient.delete(`/customers/${customerId}/contact-records/${recordId}`);
}

export {
  getCustomerList,
  getCustomer,
  createCustomer,
  createSeaCustomer,
  checkSeaCustomerDuplicateContacts,
  updateCustomer,
  updateSeaCustomer,
  updateSeaCustomerConsultation,
  getCustomerSearch,
  getCustomerShares,
  shareCustomer,
  unshareCustomer,
  releaseCustomerToSea,
  claimCustomerFromSea,
  voidCustomer,
  deleteCustomer,
  addCustomerContact,
  updateCustomerContact,
  removeCustomerContact,
  addCustomerContactRecord,
  updateCustomerContactRecord,
  removeCustomerContactRecord,
};
